Restore thread starvation tests

lorban opened the pull request:
`testReadStarvation` are broken in ee9, ee10 and core.

They uncovered a problem with the core API not always correctly using `InvocationType`s and using `CompletableFuture` as a return type for async calls.

lorban committed:
#12214 restore ee9 and ee10 thread starvation tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#12214 implement todos

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#12214 add read starvation test to ee* envs

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Fixed EE10 read starvation

gregw commented:
@lorban I've pushed a fixed for the read side of servlets in ee10. Thoughts?

gregw commented:
@lorban @sbordet we could extend CompletableFuture to an InvocableCompletableFuture that either:
 + intercepts calls to methods like `andThen` which pass in functional references.  If any such method is called, then the invocation type will be blocking, otherwise it is non-blocking
 + intercepts calls to methods like `andThen` and looks at the invocation type of the functional references passed, which it combines as the invocation type.  
 + intercepts calls to methods like `andThen` and redirects them to `andThenAsync`.  The invocation type can then always be non-blocking.

Note the returned completablefuture might also need to be wrapped (maybe not for the andThenAsync redirection).

lorban committed:
#12214 fix Content.Source.asString() so its unblocking task is NON_BLOCKING

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#12214 fix ee9 testReadStarvation

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
@lorban @sbordet would this help:
```java


    class NonBlockingCompletableFuture<V> extends CompletableFuture<V> implements Invocable
    {
        private final Executor _executor;

        public NonBlockingCompletableFuture(Executor executor)
        {
            _executor = executor;
        }

        @Override
        public InvocationType getInvocationType()
        {
            return InvocationType.NON_BLOCKING;
        }

        @Override
        public CompletableFuture<Void> acceptEither(CompletionStage<? extends V> other, Consumer<? super V> action)
        {
            return super.acceptEitherAsync(other, action, _executor);
        }

        @Override
        public <U> CompletableFuture<U> applyToEither(CompletionStage<? extends V> other, Function<? super V, U> fn)
        {
            return super.applyToEitherAsync(other, fn, _executor);
        }

        @Override
        public CompletableFuture<V> exceptionallyCompose(Function<Throwable, ? extends CompletionStage<V>> fn)
        {
            return super.exceptionallyComposeAsync(fn, _executor);
        }

        @Override
        public <U> CompletableFuture<U> handle(BiFunction<? super V, Throwable, ? extends U> fn)
        {
            return super.handleAsync(fn, _executor);
        }

        @Override
        public CompletableFuture<Void> runAfterBoth(CompletionStage<?> other, Runnable action)
        {
            return super.runAfterBothAsync(other, action, _executor);
        }

        @Override
        public CompletableFuture<Void> runAfterEither(CompletionStage<?> other, Runnable action)
        {
            return super.runAfterEitherAsync(other, action, _executor);
        }

        @Override
        public CompletableFuture<Void> thenAccept(Consumer<? super V> action)
        {
            return super.thenAcceptAsync(action, _executor);
        }

        @Override
        public <U> CompletableFuture<Void> thenAcceptBoth(CompletionStage<? extends U> other, BiConsumer<? super V, ? super U> action)
        {
            return super.thenAcceptBothAsync(other, action, _executor);
        }

        @Override
        public <U> CompletableFuture<U> thenApply(Function<? super V, ? extends U> fn)
        {
            return super.thenApplyAsync(fn, _executor);
        }

        @Override
        public <U, V1> CompletableFuture<V1> thenCombine(CompletionStage<? extends U> other, BiFunction<? super V, ? super U, ? extends V1> fn)
        {
            return super.thenCombineAsync(other, fn, _executor);
        }

        @Override
        public <U> CompletableFuture<U> thenCompose(Function<? super V, ? extends CompletionStage<U>> fn)
        {
            return super.thenComposeAsync(fn, _executor);
        }

        @Override
        public CompletableFuture<Void> thenRun(Runnable action)
        {
            return super.thenRunAsync(action, _executor);
        }

        @Override
        public CompletableFuture<V> whenComplete(BiConsumer<? super V, ? super Throwable> action)
        {
            return super.whenCompleteAsync(action, _executor);
        }
    }
```

sbordet commented:
I honestly would not go there.
Changing the semantic of `thenRun()` into a `thenRunAsync()` is breaking the least surprise.

I think we can just say in the javadocs to never block on the CF returned, or deprecate the API.

gregw commented:
> I honestly would not go there. Changing the semantic of `thenRun()` into a `thenRunAsync()` is breaking the least surprise.
> 
> I think we can just say in the javadocs to never block on the CF returned, or deprecate the API.

We need more than javadocs, as we need the Runnable or Callback created from the CF to be Invocable.  So how about:
```java
    /**
     * An extension of {@link java.util.concurrent.CompletableFuture} that is an {@link Invocable}.
     * The {@link InvocationType} is initially the type used in construction (default NON_BLOCKING).
     * If a non async method is called, then the invocation type of any passed function is used.
     * @param <V>
     */
    class InvocableCompletableFuture<V> extends java.util.concurrent.CompletableFuture<V> implements Invocable
    {
        private final AtomicReference<InvocationType> _invocationType = new AtomicReference<>();

        public InvocableCompletableFuture()
        {
            this(null);
        }

        public InvocableCompletableFuture(InvocationType invocationType)
        {
            _invocationType.set(Objects.requireNonNullElse(invocationType, InvocationType.NON_BLOCKING));
        }

        @Override
        public InvocationType getInvocationType()
        {
            return _invocationType.get();
        }

        @Override
        public java.util.concurrent.CompletableFuture<Void> acceptEither(CompletionStage<? extends V> other, Consumer<? super V> action)
        {
            _invocationType.set(Invocable.combine(Invocable.getInvocationType(other), Invocable.getInvocationType(action)));
            return super.acceptEither(other, action);
        }

        @Override
        public <U> java.util.concurrent.CompletableFuture<U> applyToEither(CompletionStage<? extends V> other, Function<? super V, U> fn)
        {
            _invocationType.set(Invocable.combine(Invocable.getInvocationType(other), Invocable.getInvocationType(fn)));
            return super.applyToEither(other, fn);
        }

        @Override
        public <U> java.util.concurrent.CompletableFuture<U> handle(BiFunction<? super V, Throwable, ? extends U> fn)
        {
            _invocationType.set(Invocable.getInvocationType(fn));
            return super.handle(fn);
        }

        @Override
        public java.util.concurrent.CompletableFuture<Void> runAfterBoth(CompletionStage<?> other, Runnable action)
        {
            _invocationType.set(Invocable.combine(Invocable.getInvocationType(other), Invocable.getInvocationType(action)));
            return super.runAfterBoth(other, action);
        }

        @Override
        public java.util.concurrent.CompletableFuture<Void> runAfterEither(CompletionStage<?> other, Runnable action)
        {
            _invocationType.set(Invocable.combine(Invocable.getInvocationType(other), Invocable.getInvocationType(action)));
            return super.runAfterEither(other, action);
        }

        @Override
        public java.util.concurrent.CompletableFuture<Void> thenAccept(Consumer<? super V> action)
        {
            _invocationType.set(Invocable.getInvocationType(action));
            return super.thenAccept(action);
        }

        @Override
        public <U> java.util.concurrent.CompletableFuture<Void> thenAcceptBoth(CompletionStage<? extends U> other, BiConsumer<? super V, ? super U> action)
        {
            _invocationType.set(Invocable.combine(Invocable.getInvocationType(other), Invocable.getInvocationType(action)));
            return super.thenAcceptBoth(other, action);
        }

        @Override
        public <U> java.util.concurrent.CompletableFuture<U> thenApply(Function<? super V, ? extends U> fn)
        {
            _invocationType.set(Invocable.getInvocationType(fn));
            return super.thenApply(fn);
        }

        @Override
        public <U, V1> java.util.concurrent.CompletableFuture<V1> thenCombine(CompletionStage<? extends U> other, BiFunction<? super V, ? super U, ? extends V1> fn)
        {
            _invocationType.set(Invocable.combine(Invocable.getInvocationType(other), Invocable.getInvocationType(fn)));
            return super.thenCombine(other, fn);
        }

        @Override
        public <U> java.util.concurrent.CompletableFuture<U> thenCompose(Function<? super V, ? extends CompletionStage<U>> fn)
        {
            _invocationType.set(Invocable.getInvocationType(fn));
            return super.thenCompose(fn);
        }

        @Override
        public java.util.concurrent.CompletableFuture<Void> thenRun(Runnable action)
        {
            _invocationType.set(Invocable.getInvocationType(action));
            return super.thenRun(action);
        }

        @Override
        public java.util.concurrent.CompletableFuture<V> whenComplete(BiConsumer<? super V, ? super Throwable> action)
        {
            _invocationType.set(Invocable.getInvocationType(action));
            return super.whenComplete(action);
        }
    }
```

gregw commented:
That would then be used with:
```java
public abstract class ContentSourceCompletableFuture<X> extends Invocable.InvocableCompletableFuture<X> implements Runnable
{
    private final Content.Source _content;

    public ContentSourceCompletableFuture(Content.Source content)
    {
        _content = content;
    }

    /**
     * <p>Initiates the parsing of the {@link Content.Source}.</p>
     * <p>For every valid chunk that is read, {@link #parse(Content.Chunk)}
     * is called, until a result is produced that is used to
     * complete this {@link CompletableFuture}.</p>
     * <p>Internally, this method is called multiple times to progress
     * the parsing in response to {@link Content.Source#demand(Runnable)}
     * calls.</p>
     * <p>Exceptions thrown during parsing result in this
     * {@link CompletableFuture} to be completed exceptionally.</p>
     */
    public void parse()
    {
        while (true)
        {
            Content.Chunk chunk = _content.read();
            if (chunk == null)
            {
                _content.demand(this);
                return;
            }
            if (Content.Chunk.isFailure(chunk))
            {
                if (chunk.isLast())
                {
                    completeExceptionally(chunk.getFailure());
                }
                else
                {
                    if (onTransientFailure(chunk.getFailure()))
                        continue;
                    _content.fail(chunk.getFailure());
                    completeExceptionally(chunk.getFailure());
                }
                return;
            }

            try
            {
                X x = parse(chunk);
                if (x != null)
                {
                    complete(x);
                    return;
                }
            }
            catch (Throwable failure)
            {
                completeExceptionally(failure);
                return;
            }
            finally
            {
                chunk.release();
            }

            if (chunk.isLast())
            {
                completeExceptionally(new EOFException());
                return;
            }
        }
    }

    @Override
    public void run()
    {
        parse();
    }

    /**
     * <p>Called by {@link #parse()} to parse a {@link org.eclipse.jetty.io.Content.Chunk}.</p>
     *
     * @param chunk The chunk containing content to parse. The chunk will never be {@code null} nor a
     *              {@link org.eclipse.jetty.io.Content.Chunk#isFailure(Content.Chunk) failure chunk}.
     *              If the chunk is stored away to be used later beyond the scope of this call,
     *              then implementations must call {@link Content.Chunk#retain()} and
     *              {@link Content.Chunk#release()} as appropriate.
     * @return The parsed {@code X} result instance or {@code null} if parsing is not yet complete
     * @throws Throwable If there is an error parsing
     */
    protected abstract X parse(Content.Chunk chunk) throws Throwable;

    /**
     * <p>Callback method that informs the parsing about how to handle transient failures.</p>
     *
     * @param cause A transient failure obtained by reading a {@link Content.Chunk#isLast() non-last}
     *             {@link org.eclipse.jetty.io.Content.Chunk#isFailure(Content.Chunk) failure chunk}
     * @return {@code true} if the transient failure can be ignored, {@code false} otherwise
     */
    protected boolean onTransientFailure(Throwable cause)
    {
        return false;
    }
}
```

gregw committed:
Use Invocable.InvocableCompletableFuture for ContentSourceCompletableFuture

gregw commented:
@sbordet @lorban I've used the approach above to extend Invocable.InvocableCompletableFuture for ContentSourceCompletableFuture.
I've tested this with a new form based starvation test, that fails with the current future and passes with this fix.

gregw committed:
more demand invocables

gregw commented:
> I honestly would not go there. Changing the semantic of `thenRun()` into a `thenRunAsync()` is breaking the least surprise.

 I think there is no problem changing `thenRun()` into `thenRunAsync()`, as the user of a CF has no control of what thread calls them in the first place, so the differences will be transparent to them.   I don't think anybody would be surprised.

But, the current approach doesn't need this.

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
IMHO the `CompletableFuture` API is orthogonal to the main problem, so we should not distract ourselves too much with this side problem.

That being said, I think the main problem I have with that API is that it makes it so easy to write broken code that I'm tempted to say it encourages. This simple example brings back the original problem in a non-obvious way:
```java
String s = Content.Source.asStringAsync(source, charset).thenRun(() -> { /* whatever */ }).get();
```
The above is my main motivation to say we should deprecate all our `CompletableFuture` API.

Back to the main problem, I remember a few months ago I said that we made a mistake when we designed the `Content.Source.demand()` call to take a `Runnable`: it should have taken a `Invocable.Task` instead to make it apparent that `InvocationType` is meaningful. I wonder if we could do that change in 12.1 without breaking backward compatibility, as it's source-compatible but not binary compatible?

lorban committed:
cleanup ee9 fix

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
cleanup ee10 fix

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
cleanup core fix

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove redundant test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
use inner classes for invocable tasks

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
use inner classes for invocable tasks

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented:
@lorban unfortunately it's not a compatible change: we have several occurrences of allocating a `Runnable` anonymous, and then calling `demand(this)` inside that `Runnable`.

sbordet commented:
> I think there is no problem changing thenRun() into thenRunAsync(), as the user of a CF has no control of what thread calls them in the first place, so the differences will be transparent to them. I don't think anybody would be surprised.

They will be, as `thenRunAsync()` will occupy a common pool thread, and after N of them they won't be available, so tasks won't be run.
In the other case I would block some other thread, but not the precious ones in the common pool.

Furthermore, it would not solve the problem of a double composition, e.g. `async().thenApply(...).thenRun(<JDBC>)`.

I really would not go into trying to put another smartness to fix this, but rather just document clearly the API or deprecate it.

lorban committed:
review usages of CS.demand() w.r.t InvocationType

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
I've just noticed some other complicating factor we must be careful about: a lot of places where an `Invocable` is passed to `demand()`, the `InvocationType` is pre-computed at the time `demand()` is called, not when the demand callback is to be executed.

This means if we ever decide to manipulate the `InvocationType` dynamically, we have to be extra careful that all `demand()` calls take that into account, which isn't the case now.

sbordet commented:
@gregw, also, `Promise.Completable` and `Callback.Completable` are already `Invocable`, so perhaps we just need a way to specify the `InvocationType`, or just assume it is non-blocking.

sbordet commented:
See alternative at #12406.

gregw commented:
@lorban 
> a lot of places where an `Invocable` is passed to `demand()`, the `InvocationType` is pre-computed at the time `demand()` is called, not when the demand callback is to be executed.

This is somewhat of an issue, but it is something we have elsewhere in the code (see dyanamic handler invocation type).  
Ultimately it is a race we can never win, if a passed callback is non-blocking, but can asynchronously become blocking, then we can always lose that race.    Basically if code that self described as asynchronous suddenly decides it is not, changes its self description and blocks, then we can do nothing.

It is something worth javadocing, but I do not think it is fixable.

gregw commented:
@sbordet 
> I really would not go into trying to put another smartness to fix this, but rather just document clearly the API or deprecate it.

"This fix" is not attempting this smartness we are discussing here.... but I think it is worthwhile discussing it so that we fully understand the problem and possible ways out.
 Let's not cede too much authority to the application by saying it would be surprised if it was called back by one thread or another.   The thread we use to call back a CF is entirely our choice, either before we call succeeded, or within the implementation of any CF that we create.

> > I think there is no problem changing thenRun() into thenRunAsync(), as the user of a CF has no control of what thread calls them in the first place, so the differences will be transparent to them. I don't think anybody would be surprised.
> They will be, as `thenRunAsync()` will occupy a common pool thread, and after N of them they won't be available, so tasks won't be run. In the other case I would block some other thread, but not the precious ones in the common pool.

Firstly, this uses a passed executor, so there is no common pool thread involved, only our executor.

But I do not think this changes behaviour.   If our thread pool is exhausted, currently our CF is seen as blocking and just never called, so deadlock happens always.   If we did a change based on converting non-async methods to async, then in the blocking case, we still hit the exhausted thread pool and are deadlocked.  

> Furthermore, it would not solve the problem of a double composition, e.g. `async().thenApply(...).thenRun(<JDBC>)`.

Yeah, that would require further wrapping... hence I don't like this approach.


So I'm not proposing we do this, and this PR is not doing it.     But I do believe that if we want to control the thread that does the callback, it is within our authority both as the caller and as the implementor of the CF, so there would be no surprise.

gregw commented:
> The above is my main motivation to say we should deprecate all our `CompletableFuture` API.

I think we should fix all our CF usages (it is not proving to be too difficult) and then separately consider if it is indeed the best API to use for many of our utility classes. As you say, it is orthogonal.

There are several ways we can "fix" our existing CF usage:
 1) Change our CFs to automatically determine their invocation type by checking what callbacks are installed on them.  This appears to work, but there is some risk of CFs dynamically changing their type after calling demand.
 2) Change our CFs to have a fixed NON_BLOCKING invocation type of non-blocking and just javadoc that the `andThen` style APIs should not be used.    This would "work", but would be fragile.
 3) Change our CFs to have a fixed NON_BLOCKING invocation type and throw UnsupportedOperation exception if any `andThen` style APIs were used.   This is kind of deprecating full CF support and a step towards coming up with our own API (which we could do as well).
 4) Change our CFs to have a fixed NON_BLOCKING invocation type and throw IllegalArgumentException if any `andThen` style APIs are used with BLOCKING functions etc.

My contributions to this PR implement 1).  They are not too complex, appear to be working, but are vulnerable to dynamic changes of invocation type.
2) is essentially the same as the alternate PR, as that it just changes to type and hopes for the best.

I don't mind 3) or 4) as they solve the dynamic type issue.    3) is a big neutering of the CF API and really shouts that we should not use it and do something else.  4) is less so, as it gives a way all the API can be used.

Ultimately,  we have always known that the invocation type of our callbacks is important and previously we have always dealt with it.    For some reason, we forgot that for a little while and introduced a few utility classes that didn't do the work required to ensure a good invocation type.    I think least surprise is to not radically change our API/contract and to just go back to caring about invocation type in all our utility classes and better javadoc of the necessity to do so for others that implement the API directly (we could even give warning for non invocable demand callbacks?).  

There is still a role for Blocking callbacks to demand.  It is just that they cannot be used if their only job is to wakeup an otherwise blocked thread, as that is deadlock.  If the demand callback is genuinely blocking (eg on a database), then it will just have to wait until a thread is available before it is called and then it can block on the database.  It just cannot wait if its job is to free up a thread that would go back to the pool to call itself in the first place.

gregw commented:
For our CF based utility classes, we have two distinct styles of usage:

1) in our adaption to blocking APIs (e.g.  in ServletApiRequest)  we often do:
    ```java
            Fields formFields = FormFields.from(request).get();
    ```

2) in truly async  (e.g. DelayedHandler) we do:
    ```java
            CompletableFuture<Fields> futureFormFields = FormFields.from(getRequest(), _charset);

            // if we are done already, then we are still in the scope of the original process call and can
            // process directly, otherwise we must execute a call to process as we are within a serialized
            // demand callback.
            boolean done = futureFormFields.isDone();
            futureFormFields.whenComplete(done ? this::process : this::executeProcess);
    ```

We typically do not mix these styles, other than the `DelayedHandler` is designed to go in front of the blocking `ServletHandler`, so the `whenComplete` style usage may proceed the `get()` style usage.  However, this is not really a problem as in this case, the `get()` will never block, as it is already done/complete by the time it is given an opportunity.

However, I don't think my fix here is good for the 2) usage.   The parsing, and hence demanding, is commenced within the `from` method, which is before the call to `whenComplete`.   So we cannot use calling that method to trigger switching the `demand` callback to be blocking.   We need to know the invocation type during the from call so we could just pass in:

```java
            CompletableFuture<Fields> futureFormFields = FormFields.from(getRequest(), _charset, InvocationType.BLOCKING);

            // if we are done already, then we are still in the scope of the original process call and can
            // process directly, otherwise we must execute a call to process as we are within a serialized
            // demand callback.
            boolean done = futureFormFields.isDone();
            futureFormFields.whenComplete(done ? this::process : this::executeProcess);
```

In this case the CF returned should have the `get()` methods extended to do something like:
```java
    public T get()
    {
         if (getInvocationType() != InvocationType.NON_BLOCKING && !isDone())
             throw new IllegalStateException("Cannot call get unless NON_BLOCKING or complete");
         return super.get();
    }
```

But it would be a bit strange as we'd then need to do in `ServletAPIRequest`:
```java
            Fields formFields = FormFields.from(request, InvocationType.NON_BLOCKING).get();
```
which I think is correct, but reads strangely.   It is also strange that the passed invocation type is only used if this is the first call to `from`

As I think the current "fix" is wrong.   I'll change it today to do this.

gregw committed:
Make the InvocableCompletableFuture non-dynamic

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/12.0.x/12214-restore-thread-starvation-tests

gregw committed:
Fragile linking of InvocableCompletableFutures

gregw committed:
javadoc

gregw committed:
Deprecated the CF APIs and replaced with explicit getXxx onXxx methods

gregw committed:
Deprecated the CF APIs and replaced with explicit getXxx onXxx methods

gregw commented:
The CF fix works but is counter intuitive, fragile and mind blowing when combining CFs as we do going from core Parts to servlet Parts.

So I've pushed some changes that deprecate all the CF APIs in Fields and MultiPart and replace with explicit `onXxx` and `getXxx` methods.  Internally, these use the "fixed" CF for now, but we can replace with something different once we can actually remove the CFs
This is still a WIP, but pushed so you can see it.

gregw committed:
removed more CF code

lorban committed:
replace FutureCallback and FuturePromise usages with Blocker.* instead

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented:
@gregw I have not looked yet, but regarding this comment:

> The thread we use to call back a CF is entirely our choice, either before we call succeeded, or within the implementation of any CF that we create.

That is true, but what I would like to see is actually the following:

* Use internally our `Promise`, `Callback`, `Blocker`, etc. that specify the `InvocationType` (likely NON_BLOCKING) until the operation is finished.
* Once the operation is finished, we can complete the `CompletableFuture` using `ThreadPool.executeImmediately()`, or similar.

In this way, the semantic of `CompletableFuture` is preserved, and we do not need any special subclass of `CompletableFuture` with overrides that would need to be reviewed when the JDK adds new methods to it.

Then, if the application blocks on the `CompletableFuture`, it's the application choice, and it won't block a selector thread.

I'll review as soon as I am back.

lorban committed:
add missing getInvocationType() returning NON_BLOCKING

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
@sbordet I think we are agreed that ultimately we want to remove CF for our implementation and API.   However, currently it  is in our API (although deprecated by this PR), so we do need to fix it.   Hence I think we should have the extended CF (also deprecated) as an interim fix until we can remove the CFs entirely.

So the key things we need to do in this PR is:
 1) fix/work around the immediate problem in our usage of utility classes (mostly done)
 2) Come up with a new API that will allow others to use these classes without using CFs (done - but needs work see below).

As we are close to release time, can you focus on reviewing 2), as new API is forever.     

I like the general style of the new API in this PR: `getParts(request)` for blocking and `onParts(request, consumer, consumer)` for async.   But the arguments of the `onParts` methods need some consideration.
Currently they are like:
```java
    void onParts(Request request, BiConsumer<Parts, Throwable> immediate, InvocableBiConsumer<Parts, Throwable> future);
    default void onParts(Request request, InvocableBiConsumer<Parts, Throwable> future) 
    { onParts(request, future, future); }
```

Thoughts about this are:
 + we do need the immediate and future arguments as often we act a bit differently if the parts are being returned in the calling thread (i.e. we don't execute)... although perhaps there is some other way to signal that. 
 + we have a single consumer convenience method for when they are both the same.
 + Using the  raw `BiConsumer` and the `InvocableBiConsumer` makes it clear which one needs to have an InvocationType. Better than javadoc.
 + `BiConsumer` is a single method interface, so lambdas can still be used, although less so with the `InvocableBiConsumer`
 
An alternative would be:

```java
    void onParts(Request request, Promise<Parts> immediate, Promise<Parts> future);
    default void onParts(Request request, Promise<Parts> future) 
    { onParts(request, future, future); }
```

Thoughts:
 + This reads a bit better
 + but we lose the distinction that it is the invocation type of the future promise that matters
 + harder to use with lambdas
 + it is an existing interface.

gregw commented:
@lorban thoughts on the API comment above?

lorban commented:
I like `Promise<Parts>` because:

 - this PR adds a `Blocker.Promise` so it's trivial to implement the blocking API with the async one
 - `Promise` is already part of our async API

but I do not like that it is not `Invocable`, but I think we can make it extend that interface without problem.

I don't mind that this API is not as fluent as `CompletableFuture`: you can't use lambdas and you cannot chain actions, this isn't so bad as this is already the case with `Callback` and we don't mind it.

I don't like `InvocableBiConsumer` as it really is equivalent to `Promise` and would IMHO just add confusions about when we would use one over the other. But maybe `BiConsumer<Parts,Throwable>` could make sense to clearly show in the API what needs an invocation type. Or maybe use `Promise<Part>` and introduce `Promise.Invocable<Part>`?

Oh, and there are other places where we expose `CompletableFuture` in our API, like in the `Content.Source` static helpers. Those should be fixed and deprecated too.

gregw committed:
Converted new API on FormFields to use Promise

gregw commented:
@sbordet @lorban I've converrted the FormFields class to use `Promise` on the new API and it works fine.    Can you review that as a priority and if OK I'll convert the MultiPart classes as well.

gregw committed:
Converted new API on Multipart to use Promise

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java:
```diff
@@ -295,9 +403,47 @@ public Parser(String boundary, MultiPartCompliance multiPartCompliance, Complian
             parser = new MultiPart.Parser(Objects.requireNonNull(boundary), compliance, listener);
         }
 
+        public void parse(Content.Source content, Promise<Parts> immediate, Invocable.InvocablePromise<Parts> future)
```
Looking at this implementation (and the future ones we could have) I wonder what's the usefulness of `Promise<Parts> immediate` as figuring out if that one should be called or the future one is always going to be racy?

gregw replied:
It is a little bit racy, but it is essentially an optimization to avoid unnecessary executes, so if we lose the race, we typically just execute instead of calling directly when the results are available.

gregw replied:
@lorban also note that if we lose the race, the invocation type of the future promise would have been used for the demand callback, so we are good.  That is why we need both the immediate and future promises, as whilst we attempt the read and parse, we are using the future promises invocation type, but if we manage to completely read the content, then we can call the immediate promise as we were not actually invoked during the reading/parsing and are still in the original calling thread.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/FormFields.java:
```diff
@@ -82,17 +84,165 @@ public static Charset getFormEncodedCharset(Request request)
+        onFields(from(request, invocationType, request, charset, maxFields, maxLength), immediate, future);
+    }
+
+    private static void onFields(CompletableFuture<Fields> futureFields, Promise<Fields> immediate, InvocablePromise<Fields> future)
```
I fail to see the point of the `Promise<Fields> immediate` argument. Mind to clarify?

gregw replied:
We frequently have slightly different behaviour if the results are immediately available or not.   
If they are immediately available and we are in a handling thread, then we can just call the next handler without executing. 
But if they are not immediately available, then invocation type matters as we will be calling demand and in many cases we will execute once the results are available.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/Invocable.java:
```diff
@@ -85,6 +98,123 @@ interface Callable extends Invocable
+     * with {@link java.util.concurrent.CompletableFuture}.
+     * @param <R> The result type
+     */
+    interface InvocablePromise<R> extends Invocable, Promise<R>, BiConsumer<R, Throwable>
```
I think this interface and the following 2 `from()` static helpers should be in the `Promise` interface instead.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/EagerFormHandler.java:
```diff
@@ -53,31 +52,74 @@ public boolean handle(Request request, org.eclipse.jetty.server.Response respons
+    protected boolean handleFormFields(Request request, org.eclipse.jetty.server.Response response, Callback callback)
+    {
+        Request.Handler handler = super::handle;
+        Promise<Fields> onFields = new Promise<>()
```
You could directly create an `InvocablePromise` here as that would save the following call to `Invocable.from()` since `FormFields.onFields()`' 2nd parameter does not care about the invocation type.

gregw replied:
But I need two different ones.  One that is non-blocking and executes, the other that could block and does not execute

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/EagerFormHandler.java:
```diff
@@ -53,31 +52,74 @@ public boolean handle(Request request, org.eclipse.jetty.server.Response respons
+    protected boolean handleMultiPartFormData(Request request, String contentType, org.eclipse.jetty.server.Response response, Callback callback)
+    {
+        Request.Handler handler = super::handle;
+        Promise<ServletMultiPartFormData.Parts> onParts = new Promise<>()
```
Ditto.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/EagerFormHandler.java:
```diff
@@ -53,31 +52,74 @@ public boolean handle(Request request, org.eclipse.jetty.server.Response respons
+        };
+
+        InvocablePromise<Fields> executeOnFields = Invocable.from(request.getContext(), onFields);
+        FormFields.onFields(request, onFields, executeOnFields);
```
Since parameters 2 and 3 do exactly the same thing, I'm inclined to believe they should be merged a single argument.

gregw replied:
@lorban but they do not.  the onFields is not executed and probably blocks.  The executedOnFields executes a call to onFields and does not block!

sbordet replied:
I'm with @lorban here: this is an implementation detail and we should not provide a user API with 2 promises for the same result.
Before, we did not care.
We are deprecating the API, so an extra dispatch on a deprecated API should not be a problem.
We provide a new API based on Promise to `FormFields`.

This new API in `FormFields` with 2 promises is only used here.
We can easily replace it with a CountDownLatch here, and avoid the new API in `FormFields`.

gregw replied:
@sbordet @lorban The need for this feature is not for a deprecated API.  In fact it is for the most likely hot path when we are using handlers like `EagerFormHandler` or `DelayedHandler`.   We want to be able to load the content asynchronously, but in many (most?) cases, the content will come with the headers and already be there.  Thus we do not want to dispatch on the most common hot path case.

If you don't like the two Promise API, then we need another API that will allow the same behaviour.   We cannot have it that adding the `EagerFormHandler` or `DelayedHandler` slows down the best case requests.

An alternative API would be to pass a single Promise and an Executor that is used only if the result is not immediately available.   That is a little special purpose, but it is the only use-case we have for the two promises.

sbordet replied:
@gregw this is orthogonal to this PR. Before we did not have such mechanism.
We do not need an API, it's just an implementation detail for only one Handler.

gregw replied:
It is not orthogonal to this PR. We've always had this behaviour and it was provided by CF. If we hide/remove CF then we need an API to the handlers that need it can do it. There is no easy way for a handler to do this without API

gregw replied:
@sbordet @lorban I've found a way to only have a single Promise. But it moves the future execution logic into the onFields style methods.

lorban reviewed (commented):
I like the static `getXyz()` blocking helpers, but I'm questioning the usefulness of having 2 promises for the `onXyz()` async helpers, otherwise the async API LGTM.

And IMHO `InvocablePromise` should be `Promise.Invocable` instead.

lorban committed:
deprecate CompletableFuture API

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Rename InvocablePromise to Promise.Invocable

gregw commented:
@sbordet @lorban I'm liking the new `getXxx()` and `onXxx(Promise)` APIs.

So my mind is turning more towards the implementation.  Currently we are still using a `CF<Xxx>` internally, as it is a useful object  to add as a request attribute, so that we can make multiple calls to `getXxx()` and `onXxx(Promise)`  idempotent.   E.g. a DelayedHandler or EagerHandler can first call the `onXxx(Promise)` to read the content asynchronously before dispatching to a blocking handler/servlet that uses `getXxx()` that finds the `CF<Xxx>` as an attribute and does a `get()` from it, which will not block.  But if the async handler is not in front, that same `get()` will block whilst the content is read in blocking mode.

But I'm wondering if we need to be so idempotent.   Will we ever get a call to `getXxx()` and `onXxx(Promise)` whilst a previous one is pending?   I don't think so as the first blocks and the second only dispatches once the content is read.    So the attribute could just be the `Xxx` put there at completion, rather than a `CF<Xxx>` put there at commencement.    The danger is if we get concurrent calls to `getXxx()` and `onXxx(Promise)`, but I just don't see how that could/should happen. 

Note that the parsers themselves are still written in terms of `CF`, so there is more work to do to remove all CF usage.  But I've deprecated some of those APIs (But FormFields is problematic and will need a breaking change in 12.1 ).

I'm happy if this PR fixes the bug ugly, but introduces a new good API... and then clean up the implementation over the next few releases (probably only in 12.1).

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Request.java:
```diff
@@ -513,9 +514,19 @@ default Transport getTransport()
      * @return whether the abort succeeded
+     * @deprecated use {@link #abort(Throwable, Promise)} instead
      */
+    @Deprecated(forRemoval = true, since = "12.0.15")
```
Why was this changed?

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java:
```diff
@@ -188,6 +188,7 @@ interface AsyncContentListener extends ContentSourceListener
         default void onContentSource(Response response, Content.Source contentSource)
         {
             Content.Chunk chunk = contentSource.read();
+            // demandCallback eventually calls onContent() which calls end-user code, so its InvocationType must be BLOCKING.
```
I'd remove the comment, it is unnecessary, as it is the default: all demand callbacks are BLOCKING unless otherwise explicitly coded otherwise.

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpSender.java:
```diff
@@ -543,7 +544,7 @@ protected Action process() throws Throwable
                     demanded = true;
                     assert content != null;
-                    content.demand(this::succeeded);
+                    content.demand(Invocable.from(getInvocationType(), this::succeeded));
```
I don't think this is right, as `succeeded()` calls user code.

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/ResponseListeners.java:
```diff
@@ -502,6 +503,20 @@ private void onDemandCallback()
             chunk.release();
         }
 
+        private Invocable.InvocationType deriveOriginalContentSourcesInvocationType()
```
Rename to just `getInvocationType()`?

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/ResponseListeners.java:
```diff
@@ -641,6 +656,11 @@ private void onDemandCallback()
                 }
             }
 
+            private Invocable.InvocationType getDemandCallbackInvocationType()
```
Just `getInvocationType()`?

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPart.java:
```diff
@@ -887,6 +882,32 @@ private enum State
             FIRST, MIDDLE, HEADERS, CONTENT, COMPLETE
         }
+
+        private class DemandInvocableTask implements Invocable.Task
```
Rename to `DemandTask` and make it extend `Invocable.ReadyTask`, so the `InvocationType` is not computed every time.

gregw replied:
Oh! when did `ReadyTask` sneak into the code base.... that is a really strange name that means nothing to me.
It is more readable to implement Invocable.Task, so perhaps just add a final field to compute the invocation type only once.

Could we deprecate the public `ReadyTask` and replace it with an `Task Invocable.from(Runnable, InvocationType)`?

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Promise.java:
```diff
@@ -219,4 +221,86 @@ public Promise<W> unwrap()
+     * with {@link CompletableFuture}.
+     * @param <R> The result type
+     */
+    interface Invocable<R> extends org.eclipse.jetty.util.thread.Invocable, Promise<R>, BiConsumer<R, Throwable>
```
Why not make Promise implement Invocable directly, like Callback does?

gregw replied:
Perhaps, although doing it this way puts something in the signature to say when the invocation type of a Promise is significant or not.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Promise.java:
```diff
@@ -219,4 +221,86 @@ public Promise<W> unwrap()
+    interface Invocable<R> extends org.eclipse.jetty.util.thread.Invocable, Promise<R>, BiConsumer<R, Throwable>
+    {
+        @Override
+        default void accept(R result, Throwable error)
```
Never used, remove.

gregw replied:
It is used

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java:
```diff
@@ -295,9 +402,47 @@ public Parser(String boundary, MultiPartCompliance multiPartCompliance, Complian
             parser = new MultiPart.Parser(Objects.requireNonNull(boundary), compliance, listener);
         }
 
+        public void parse(Content.Source content, Promise<Parts> immediate, Promise.Invocable<Parts> future)
```
Not 2 promises for the same result!

It is just an implementation detail that we do not want to surface to user APIs.

gregw replied:
Resolving here and we can just discuss API in FormFields

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ChunkAccumulator.java:
```diff
@@ -222,6 +223,16 @@ public void run()
+        @Override
+        public InvocationType getInvocationType()
+        {
+            return InvocationType.NON_BLOCKING;
```
That's not correct, as it may invoke user code.

gregw replied:
This is another case when we need to know how it is to be used.  If only get is called, then it is NON_BLOCKING.  In this case, I can't find any actual usage of this class??

gregw replied:
For now I have just deprecated the API

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/Invocable.java:
```diff
@@ -241,4 +263,156 @@ public InvocationType getInvocationType()
+     * @deprecated This class in only used for deprecated usages of CompletableFuture
+     */
+    @Deprecated(forRemoval = true, since = "12.0.15")
+    class InvocableCompletableFuture<V> extends java.util.concurrent.CompletableFuture<V> implements Invocable
```
I don't think we should add new APIs, throwing ISEs from CompletableFuture methods that don't normally throw, it breaks least surprise.
I'd just deprecate the relevant APIs.

gregw replied:
I think we should not care about least surprise here, as we do not want people to use this API at all.
They are obtaining a CF after passing arguments to say how they will using it... if they end up using it wrongly (and risking deadlock) then we can catch that case with an ISE.    It is a protective layer whilst we remove this usage entirely.

sbordet replied:
We just deprecate and document it. 
It is not solving the issue, just throwing an exception in the user's face, which will likely break the application anyway, so it does not really protect users.

gregw replied:
It protects users by throwing during development and testing rather than waiting until the last thread in production deadlocks.

gregw replied:
As discussed in hangout, we will keep this whilst developing this PR and then consider removing it (note it has already helped me find one coding error)

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceTransformer.java:
```diff
@@ -157,4 +159,28 @@ private Content.Chunk process(Content.Chunk rawChunk)
      */
     protected abstract Content.Chunk transform(Content.Chunk inputChunk);
+
+    private class DemandInvocableTask implements Invocable.Task
```
Just `DemandTask` and extends `ReadyTask`.

gregw replied:
I prefer as is, but with final field for InvocationType

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java:
```diff
@@ -57,7 +58,7 @@ protected Action process() throws Throwable
         if (current == null)
         {
-            source.demand(this::succeeded);
+            source.demand(Invocable.from(getInvocationType(), this::succeeded));
```
No, the nested callback might be blocking.

gregw replied:
Should the invocationType be that of the callback?

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceString.java:
```diff
@@ -71,4 +75,19 @@ private void succeed()
         }
     }
+
+    private class ConvertInvocableTask implements Invocable.Task
```
Rename to `DemandTask extends ReadyTask`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/FormFields.java:
```diff
@@ -82,17 +84,165 @@ public static Charset getFormEncodedCharset(Request request)
+     * @see #getFields(Request)
+     * @see #getFields(Request, int, int)
+     */
+    public static void onFields(Request request, Promise<Fields> immediate, Promise.Invocable<Fields> future)
```
Not 2 promises for the same result.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextRequest.java:
```diff
@@ -59,11 +59,11 @@ public Context getContext()
     }
 
-    private class OnContextDemand implements Runnable
+    private class OnContextDemandInvocableTask implements Invocable.Task
```
`DemandTask extends ReadyTask`.

gregw replied:
see above

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpInput.java:
```diff
@@ -159,6 +160,18 @@ public boolean isAsync()
         return _readListener != null;
     }
 
+    public Invocable.InvocationType getReadListenerInvocationType()
```
Just `getInvocationType()`.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -406,4 +401,26 @@ public String toString()
         }
     }
+
+    private class IsReadyInvocableTask implements Invocable.Task
```
`DemandTask extends ReadyTask`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java:
```diff
@@ -365,6 +367,28 @@ private void onPermittedContent(Permit permit)
             }
         }
+
+        private class OnContentInvocableTask implements Invocable.Task
```
`DemandTask extends ReadyTask`.

gregw replied:
see above

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannel.java:
```diff
@@ -539,13 +539,13 @@ public void handle()
                     case READ_CALLBACK:
                     {
-                        _context.run(() -> _servletContextRequest.getHttpInput().run());
+                        _context.run(_servletContextRequest.getHttpInput()::readCallback);
```
```suggestion
                        _context.run(_servletContextRequest.getHttpInput());
```

gregw replied:
I prefer the former for readability.    Besides, HttpInput is no longer a Runnable

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannel.java:
```diff
@@ -539,13 +539,13 @@ public void handle()
                     case WRITE_CALLBACK:
                     {
-                        _context.run(() -> _servletContextRequest.getHttpOutput().run());
+                        _context.run(_servletContextRequest.getHttpOutput()::writeCallback);
```
```suggestion
                        _context.run(_servletContextRequest.getHttpOutput());
```

gregw replied:
I prefer the former as it is more readable

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletMultiPartFormData.java:
```diff
@@ -51,15 +54,88 @@
+     * @param future The action to take when the Parts are available, if they are not available immediately.  The {@link org.eclipse.jetty.util.thread.Invocable.InvocationType}
+     *               of this parameter will be used as the type for any implementation calls to {@link Content.Source#demand(Runnable)}.
+     */
+    static void onParts(ServletRequest servletRequest, String contentType, Promise<Parts> immediate, Promise.Invocable<Parts> future)
```
Not 2 promises for the same result.

gregw replied:
See discussion in FormFields for resolution

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/HttpChannel.java:
```diff
@@ -1602,4 +1602,20 @@ public void dispatch() throws IOException, ServletException
         }
     }
+
+    private class NeedContentInvocableTask implements Invocable.Task
```
`DemandTask extends readyTask`.

gregw replied:
In this case we cannot use ReadyTask, as the invocation type will not be known at the time of construction.

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/HttpInput.java:
```diff
@@ -365,6 +366,18 @@ public int available()
         }
     }
 
+    public Invocable.InvocationType getReadListenerInvocationType()
```
Just `getInvocationType()`.

sbordet reviewed (changes requested):
This PR does not take into account that `SerializedInvoker` currently makes every `Runnable` passed in a BLOCKING one.
Since demand callbacks are always run through a `SerializedInvoker`, specifying NON_BLOCKING demand callbacks would not work as expected.
We need to test this, and possibly find a solution for `SerializedInvoker`.

sbordet commented:
For `SerializedInvoker` an experiment I made is the following:

In `Link.run()` we remember the `InvocationType` of the running task.
If, after running a task, it is detected that another task should be run, we check the `InvocationType` of the new task against the previous task.
Only in the case prev=NON_BLOCKING and next=BLOCKING, we wrap the next task, execute it and return.
The wrapper will resume `Link.run()` to look for a next-next task.
This preserves the task serialization, although not anymore _same_ thread-ness (although guarantees _single_ thread-ness).

gregw committed:
Alternative to only have a single Promise

gregw committed:
Alternative to only have a single Promise

gregw committed:
Alternative to only have a single Promise

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/FormFields.java:
```diff
@@ -82,17 +83,126 @@ public static Charset getFormEncodedCharset(Request request)
+        onFields(request, getFormEncodedCharset(request), promise);
+    }
+
+    /**
+     * Asynchronously read and parse FormFields from a {@link Request}.
+     * <p>
+     * Calls to {@code onFields} and {@code getFields} methods are idempotent, and
+     * can be called multiple times, with subsequent calls returning the results of the first call.
+     * @param request The request to get or read the Fields from
+     * @param charset The {@link Charset} of the request content, if previously extracted.
+     * @param promise The action to take when the FormFields are available.
+     *                If the {@link InvocationType} of the promise is not {@link InvocationType#NON_BLOCKING},
+     *                then any calls to the promise may be executed via the {@link Request#getContext()} so that
+     *                the implementation can pass a {@link InvocationType#NON_BLOCKING} {@link Runnable} to
+     *                {@link Content.Source#demand(Runnable)}.  If the fields are available immediately, then the promise
+     *                will always be called directly from within the onFields call.
+     * @see #onFields(Request, Charset, Promise.Invocable)
+     * @see #getFields(Request)
+     * @see #getFields(Request, int, int)
+     */
+    public static void onFields(Request request, Charset charset, Promise.Invocable<Fields> promise)
+    {
+        InvocationType invocationType = promise.getInvocationType();
+        int maxFields = getContextAttribute(request.getContext(), FormFields.MAX_FIELDS_ATTRIBUTE, FormFields.MAX_FIELDS_DEFAULT);
+        int maxLength = getContextAttribute(request.getContext(), FormFields.MAX_LENGTH_ATTRIBUTE, FormFields.MAX_LENGTH_DEFAULT);
+        CompletableFuture<Fields> futureFields = from(request, InvocationType.NON_BLOCKING, request, charset, maxFields, maxLength);
+
+        if (futureFields.isDone() || invocationType == InvocationType.NON_BLOCKING)
+            futureFields.whenComplete(promise);
+        else
+            futureFields.whenComplete(Promise.from(request.getContext(), promise));
+    }
```
@sbordet @lorban I've replaced the two Promise version with this API.  However, it does an execute internally to avoid BLOCKING callbacks.

I've updated the ThreadStarvationTest to include DelayedHandler to test this.

gregw committed:
Alternative to only have a single Promise

gregw committed:
Alternative to only have a single Promise

gregw committed:
Alternative to only have a single Promise

gregw committed:
Alternative alternative to only have a single Promise

gregw committed:
Alternative alternative to only have a single Promise

sbordet commented:
@gregw the `SerializedInvoker` issue we discussed is as follows:

* The application demand callback is registered as `HttpChannelState._onContentAvailable`, and the invocation type checked against it directly.
* However, when it needs to be run because content is available, method `HttpChannelState.onContentAvailable()` is called, that wraps the application demand callback in a `SerializedInvoker.Link`, whose `InvocationType` is always `BLOCKING`.
* The `Runnable` returned from `HttpChannelState.onContentAvailable()` is then:
  * For HTTP/1.1, run immediately
  * For HTTP/2 and HTTP/3, it is given back to the `AdaptiveExecutionStrategy`, which will consume it accordingly to its invocation type, i.e. dispatch it because it is `BLOCKING`.

Example: request arrives, but no content yet; application registers a non-blocking demand callback; then:
* For HTTP/1.1, a non-blocking callback is registered to NIO; when NIO wakes up, runs it directly; the application demand callback is wrapped in a `Link`, but then it is run immediately despite being `BLOCKING` (`HttpConnection.DemandContentCallback`).
* For HTTP/2 and HTTP/3, when data arrives, it is queued into the stream, which checks whether there is demand; the application demand callback is wrapped in a `Link`, and then given back to the `AdaptiveExecutionStrategy` that it runs it as `BLOCKING` which is going to be inefficient (`HttpStreamOverHTTP2.onDataAvailable()`).

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartConfig.java:
```diff
@@ -17,12 +17,14 @@
  * The Configuration needed to parse multipart/form-data.
- * @see MultiPartFormData#from(Content.Source, Attributes, String, MultiPartConfig)
+ * @see MultiPartFormData#getParts(Content.Source, Attributes, String, MultiPartConfig)
+ * @see MultiPartFormData#onParts(Content.Source, Attributes, String, MultiPartConfig, Promise, Promise.Invocable)
```
Update this, we don't have anymore 2 promises.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java:
```diff
@@ -126,35 +126,11 @@ public static MultiPartFormData.Parts getParts(Content.Source content, Attribute
-        {
-            futureParts.whenComplete(future);
-        }
+        from(content, promise.getInvocationType(), attributes, contentType, config).whenComplete(promise);
```
This can be written the other way around, leveraging existing methods that have been introduced a while ago.

```suggestion
        promise.completeWith(from(content, promise.getInvocationType(), attributes, contentType, config));
```

In this way, we can simplify `Promise.Invocable` and eventually just make `Promise` implement `Invocable` without the need of an extra class.

gregw replied:
However, `promise.completeWith` is an extra allocation as it creates a lambda on every call.    Having a Promise that is also a BiConsumer is a reasonable optimization that avoids this allocation.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/FormFields.java:
```diff
@@ -169,71 +164,17 @@ public static void onFields(Request request, Promise.Invocable<Fields> future)
-        {
-            futureFields.whenComplete(future);
-        }
+        from(request, promise.getInvocationType(), request, charset, maxFields, maxLength).whenComplete(promise);
```
```suggestion
        promise.completeWith(from(request, promise.getInvocationType(), request, charset, maxFields, maxLength));
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DelayedHandler.java:
```diff
@@ -250,7 +251,9 @@ public UntilFormDelayedProcess(Handler handler, Request wrapped, Response respon
         {
-            Promise<Fields> onFields = new Promise<>()
+            InvocationType invocationType = getHandler().getInvocationType();
+            AtomicBoolean immediate = new AtomicBoolean(true);
```
I'd rename the variable to `done` and invert the true/false, but it's a small nit.

gregw replied:
Actually I think counting down an AtomicInteger from 2 is better.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/EagerFormHandler.java:
```diff
@@ -61,8 +62,10 @@ public boolean handle(Request request, org.eclipse.jetty.server.Response respons
-        Promise<Fields> onFields = new Promise<>()
+        Request.Handler handler = getHandler();
+        InvocationType invocationType = handler.getInvocationType();
+        AtomicBoolean immediate = new AtomicBoolean(true);
```
I'd rename the variable to `done` and invert the true/false, but it's a small nit.

gregw replied:
counting down from an atomic integer of 2 is better

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletMultiPartFormData.java:
```diff
@@ -79,38 +78,12 @@ static Parts getParts(ServletRequest servletRequest)
-            futureParts.whenComplete(future);
-        }
+        CompletableFuture<Parts> futureParts = from(servletRequest, promise.getInvocationType(), contentType);
+        futureParts.whenComplete(promise);
```
```suggestion
        promise.completeWith(futureParts);
```

sbordet committed:
Ported good changes from #12406.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed compilation issue.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed javadoc link.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed NullPointerExceptions.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
rename immediate to done

gregw commented:
> @gregw the `SerializedInvoker` issue we discussed is as follows:

@sbordet So do you think we should write a simplified SerializedInvoker that only prevents infinite recursion but does not support queuing infinite jobs (only 0 or 1), so that it can correctly report an invocation type?

sbordet committed:
Modified SerializedInvoker to take into account the task InvocationType.

Added ThreadStarvationTest for all transports, to check that also HTTP/2 and HTTP/3 do not starve in case of non-blocking reads.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
Improved Core ThreadStarvationTest

sbordet committed:
Fixed checkstyle.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed AsyncServletLongPollTest in ee9 and ee10 to match the current behavior in case of close.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
TODO:

* Make `Promise` implement `Invocable`.
* Remove `Promise.Invocable`
* Remove `Invocable.InvocableCompletableFuture`.
* Use `Invocable.ReadyTask` in more places where `Invocable.Task` is used.

gregw commented:
> TODO:
> 
> * Make `Promise` implement `Invocable`.

I don't think we should.   Not all Promises are Invocable, and this makes it clear in the signature when the caller has to think about InvocationType or not.

> * Remove `Promise.Invocable`

Keep it.

> * Remove `Invocable.InvocableCompletableFuture`.

I'm 50:50 on this one.  It has helped me a couple of times when I got the wrong invocation type during development.

> * Use `Invocable.ReadyTask` in more places where `Invocable.Task` is used.

I find the name `ReadyTask` meaningless and it does very little.  Directly implementing `Invocable.Task` is more readable.   Also, we end up creating Runnable lambdas just to pass into the constructor of ReadyTask, so might as well just create a nested class anyway.

gregw commented:
> > * Remove `Invocable.InvocableCompletableFuture`.
> 
> I'm 50:50 on this one. It has helped me a couple of times when I got the wrong invocation type during development.

I've got a compromise.  I've removed `InvocableCompletableFuture`, but moved its methods to `ContentSourceCompletableFuture`, which was the only use anyway.  We have not yet deprecated CSCF as we do not have an alternative, so the extra protection in using it is worthwhile I think.

gregw committed:
Removed Invocable.InvocableCompletableFuture

gregw commented:
@sbordet  Another compromise....  How about we introduce `Invocable.AbstractTask` that implements the getInvocationType, but keeps run abstract so we don't have to create a runnable lambda just to pass into the ReadyTask constructor?

sbordet committed:
Fixed SerializedInvoker.
Improved ThreadStarvationTest.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
Update from review:

 + refined more DemandTask implementations
 + Added InvocableType.runWithoutBlocking (name is WIP)

gregw committed:
Update from review

gregw committed:
Update from review

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceCompletableFuture.java:
```diff
@@ -53,25 +66,34 @@
  * }</pre>
  */
-public abstract class ContentSourceCompletableFuture<X> extends CompletableFuture<X>
+public abstract class ContentSourceCompletableFuture<X> extends CompletableFuture<X> implements Runnable, Invocable
```
Do not implement `Runnable` and `Invocable` but solely `Invocable.Task` instead.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextRequest.java:
```diff
@@ -37,8 +37,7 @@ protected ContextRequest(ContextHandler.ScopedContext context, Request request)
     {
-        // inner class used instead of lambda for clarity in stack traces
-        super.demand(new OnContextDemand(demandCallback));
+        super.demand(new DemandTask(demandCallback));
```
The comment should be preserved IMHO.

lorban commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/DelayedHandlerTest.java:
```diff
@@ -503,4 +508,72 @@ public boolean handle(Request request, Response response, Callback callback) thr
+            @Override
+            public boolean handle(Request request, Response response, Callback callback) throws Exception
+            {
+                CompletableFuture<MultiPartFormData.Parts> future = MultiPartFormData.get(request);
```
Shouldn't we test the new API here?

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/SerializedInvoker.java:
```diff
@@ -235,27 +243,43 @@ Link next()
+
+                Runnable task = link.getTask();
+                InvocationType currentInvocationType = link.getInvocationType();
+                if (currentInvocationType == InvocationType.BLOCKING && firstInvocationType != InvocationType.BLOCKING)
```
This new logic deserves a test.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/AsyncServletLongPollTest.java:
```diff
@@ -202,12 +183,18 @@ public void destroy()
             await().atMost(5, TimeUnit.SECONDS).until(asyncContextRef::get, Matchers.notNullValue());
 
+            // Wait for the request on the server to become idle.
+            Thread.sleep(1000);
```
I'm sure awaitility can be used to replace this `sleep`, possibly by waiting for the thread that executed the handler to have its status changed to WAITING.

lorban commented on jetty-ee9/jetty-ee9-servlet/src/test/java/org/eclipse/jetty/ee9/servlet/AsyncServletLongPollTest.java:
```diff
@@ -201,12 +182,18 @@ public void destroy()
             await().atMost(5, TimeUnit.SECONDS).until(asyncContextRef::get, Matchers.notNullValue());
 
+            // Wait for the request on the server to become idle.
+            Thread.sleep(1000);
```
Can probably be replaced by awaitility.

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
replace Thread.sleep() with awaitility

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Update from review- use new APIs

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/Invocable.java:
```diff
@@ -85,36 +120,56 @@ interface Callable extends Invocable
         void call() throws Exception;
     }
 
+    abstract class AbstractTask implements Task
```
Not `Invocable.Task.Abstract` like we did in `Handler`?

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/Invocable.java:
```diff
@@ -69,6 +83,27 @@ enum InvocationType
+         * @param task The task to run
+         * @param executor The executor to use if necessary
+         */
+        public abstract void runWithoutBlocking(Runnable task, Executor executor);
```
The equivalent method in AES is called `consume()`, which is shorter than `runWithoutBlocking()`.
Not a biggie though.

gregw replied:
Whilst I don't really like `runWithoutBlocking`, it does not have the context of AES to tell us exactly what it is doing... so I think the explicit name is best.

sbordet reviewed (changes requested):
If you want to tackle the 2 renamings will be good, otherwise LGTM.

gregw committed:
Update from review- rename AbstractTask to Task.Abstract

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/12.0.x/12214-restore-thread-starvation-tests
